web/source/kmwosk.js

```diff
@@ -2490,6 +2729,165 @@ if(!window['tavultesoft']['keymanweb']['initialized']) {
     }
 
     /**
+     * Converts the legacy BK property from pre 10.0 into the KLS keyboard layer spec format,
+     * sparsifying it as possible to pre-emptively check invalid layers.
+     * 
+     * @param   {Array}   BK      keyboard object (as loaded)
+     * @return  {Object}
+     */
+    osk.processLegacyDefinitions = function(BK) {
+      //['default','shift','ctrl','shiftctrl','alt','shiftalt','ctrlalt','shiftctrlalt'];
+      var idList=osk.generateLayerIds(false); // Non-chiral.
+
+      var KLS = {};
+
+      // The old default:  eight auto-managed layers...
+      for(var n=0; n<idList.length; n++) {
+        var id = idList[n], arr = [], valid = false;
+
+        // ... with keycode mappings in blocks of 65.
+        for(var k=0; k < 65; k++) {
+          var index = k + 65 * n;
+          arr.push(BK[index]);
+
+          // The entry for K_SPACE's keycode tends to hold ' ' instead of '', which causes
+          // the whole layer to be treated as 'valid' if not included in the conditional.
+          if(index < BK.length && BK[index] != '' && k != dfltCodes.indexOf('K_SPACE')) {
+            valid = true;
+          }
+        }
+
+        if(valid) {
+          KLS[id] = arr;
+        }
+      }
+
+      // There must always be at least a plain 'default' layer.
+      if(typeof KLS['default'] == 'undefined' || ! KLS['default']) {
+        KLS['default'] = Array.from(''.repeat(65));
```

Comment:
What does ` Array.from(''.repeat(65))` do? AFAICT it gives an empty array

Reply:
Exactly.  An array of 65 empty strings, corresponding to one empty layer of 65 keys.

Reply:
Not quite. The expansion:
````
var foo=''.repeat(65);  // foo there becomes 65*'' == ''.
Array.from(foo);        // result is therefore empty array
````

Sadly also `String.prototype.repeat` is [not supported in IE](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/repeat#Browser_compatibility). Polyfill is probably overkill for this single use.

Reply:
And to add insult to injury, [IE doesn't support `Array.from`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from#Browser_compatibility) either. Sorry!

Reply:
I stand corrected.  I figured out why the code still works regardless and pointed it toward that instead for now, with a comment.